### PR TITLE
break up the test tarball

### PR DESCRIPTION
**Pull request by ixdy**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: implements [sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball](https://github.com/kubernetes/enhancements/blob/master/keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md):
- produce `kubernetes-test-portable.tar.gz` and `kubernetes-test-{OS}-{ARCH}.tar.gz`
- continue to produce `kubernetes-test.tar.gz` with an embedded `DEPRECATION_NOTICE`
  - this can be disabled in the make-based build system by setting `KUBE_BUILD_MONDO_TEST_TARBALL=n`
- update `cluster/get-kube-binaries.sh` to prefer downloading the split test tarballs
- also clean up some dead code in `get-kube-binaries.sh`

x-ref https://github.com/kubernetes/enhancements/issues/714

**Does this PR introduce a user-facing change?**:

```release-note
Split up the mondo `kubernetes-test` tarball into `kubernetes-test-portable` and `kubernetes-test-{OS}-{ARCH}` tarballs.
```

/sig testing release
/priority important-soon

### CODE REVIEW

**Comment by ixdy** [comment]
I think this is basically ready for review now. The bazel stuff somewhat conflicts with https://github.com/kubernetes/kubernetes/pull/73930, so I'll clean up the mess in whichever PR loses the race.

cc @spiffxp @tpepper @kubernetes/bash-firefighters @akutz @amwat

File: build/lib/release.sh
Comment (akutz) [inline build/lib/release.sh]: I'm just marveling at the fact that you can have colons in a function name in BASH. I had no idea.

**Comment by ixdy** [comment]
I'm not sure who to assign here... I feel bad for overburdening the same folks each time.

/assign @BenTheElder @cblecker @fejta @spiffxp

**Comment by ixdy** [comment]
also cc the release-engineering team:
@aleksandra-malinowska @calebamiles @mikedanese @tpepper

File: build/lib/release.sh
Comment (fejta) [inline build/lib/release.sh]: What does this `#` mean?

**Review by fejta — COMMENTED** [review]
/lgtm
/hold

File: build/lib/release.sh
Comment (akutz) [reply build/lib/release.sh]: Hi @fejta,

I mean, it _looks_ like string manipulation, but dang it's nasty. 

From https://www.tldp.org/LDP/abs/html/arrays.html:
> ```
> string=abcABC123ABCabc
> echo ${string[@]}               # abcABC123ABCabc
> ```

And from http://tldp.org/LDP/abs/html/parameter-substitution.html:

> `${var#Pattern}` Remove from $var the shortest part of $Pattern that matches the front end of $var.

I read it as removing the string evaluated from `/${LOCAL_OUTPUT_BINPATH}/${platform}/` from all of the elements in `${KUBE_TEST_SERVER_BINARIES}`.

I suppose it makes sense if they're trying to use the `cp` command's ability to copy multiple source files into a single destination directory:

```
cp [FLAGS] source_file ... target_directory
```

Maybe trying to remove all but the file name for each element of `${KUBE_TEST_SERVER_BINARIES}` in order to copy them into `"${release_stage}/test/bin/"`?

That's my best guess without spending more time looking at it.

Comment (akutz) [reply build/lib/release.sh]: Hi @fejta,

From the same [arrays link](https://www.tldp.org/LDP/abs/html/arrays.html) as above:

> ```
> # Substring Removal
> 
> # Removes shortest match from front of string(s).
> 
> echo ${arrayZ[@]#f*r}   # one two three five five
> #               ^       # Applied to all elements of the array.
>                         # Matches "four" and removes it.
> ```

This seems to be the same pattern being used and that I sussed out above. Still, with so many variables in play, I haven't traced out what the actual result is.

Comment (ixdy) [reply build/lib/release.sh]: @akutz was on the right path, but I think this is actually a different operator - `/#`, not `#`.

From the [arrays link](https://www.tldp.org/LDP/abs/html/arrays.html):
```
arrayZ=( one two three four five five )

# Substring Replacement

# Replace front-end occurrences of substring.
echo ${arrayZ[@]/#fi/XY}    # one two three four XYve XYve
#                ^          # Applied to all elements of the array.
```

Because the match string is empty (there's no string after `#` before the next `/`), we end up just prepending the following string. (It's extra confusing because we have `/`s in the "replacement" string.)

It's documented in a few other places it's used in this file, e.g. 
https://github.com/kubernetes/kubernetes/blob/61b676aae1e80673251fbbad3712d0e722b11f0d/build/lib/release.sh#L137-L141

I can add a similar comment for this usage.

(Have I mentioned that the bash side of this PR was far more difficult than the bazel side?)

Comment (akutz) [reply build/lib/release.sh]: Thank you @ixdy. I didn't realize that `/#` behaved differently. I don't tend to use Bash string manipulation at all, let alone Bash arrays. I usually stick with POSIX-compliant rules and defer to `sed` for replacements.

maxresdefault

Comment (ixdy) [reply build/lib/release.sh]: yeah, I blame @jbeda :) https://github.com/kubernetes/kubernetes/pull/1324

Comment (ixdy) [reply build/lib/release.sh]: definitely a cool trick, though.
https://github.com/kubernetes/kubernetes/pull/1324#discussion_r17577258

**Review by fejta — COMMENTED** [review]
/lgtm

File: build/lib/release.sh
Comment (BenTheElder) [reply build/lib/release.sh]: There's some really amazing bash in this project. And also plenty of horrifying stuff #72956 😛

File: build/lib/release.sh
Comment (BenTheElder) [inline build/lib/release.sh]: can we log some warning about the deprecation / maybe link to the KEP?

Comment (BenTheElder) [reply build/lib/release.sh]: I see that we put one in the tarball alone, but I think a lot of people never inspect these manually, whereas they might see the build output.

Comment (ixdy) [reply build/lib/release.sh]: seems kinda spammy in the build logs. maybe when the default is switched to N?

Comment (BenTheElder) [reply build/lib/release.sh]:
```suggestion
    kube::log::status "Building tarball: test mondo (deprecated)"
```
?

Comment (ixdy) [reply build/lib/release.sh]: done (and also referred to the KEP, though I wish there was a shorthand for that)

File: cluster/get-kube-binaries.sh
Comment (spiffxp) [inline cluster/get-kube-binaries.sh]: who says bash can't do sets of tuples

**Review by spiffxp — APPROVED** [review]
/lgtm

File: cluster/get-kube-binaries.sh
Comment (ixdy) [reply cluster/get-kube-binaries.sh]: I could probably use `IFS="/" read` here instead, but I felt this was more readable.

Comment (BenTheElder) [reply cluster/get-kube-binaries.sh]: who says indeed? :P

File: build/lib/release.sh
Comment (BenTheElder) [reply build/lib/release.sh]: KEP shortlinks would be awesome...

File: build/lib/release.sh
Comment (BenTheElder) [inline build/lib/release.sh]: thanks, FWIW you can get a sorta short-ish link to the file:
https://git.k8s.io/enhancements/keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

**Comment by ixdy** [comment]
/hold cancel
:shipit:

**Pull request merged**
